libraries/AP_Devo_Telem/AP_Devo_Telem.cpp

```diff
@@ -0,0 +1,133 @@
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/* 
+   DEVO Telemetry library
+*/
+#include "AP_Devo_Telem.h"
+extern const AP_HAL::HAL& hal;
+
+DevoMPacket AP_DEVO_Telem::devoPacket = {DEVOM_SYNC_BYTE, 0};
+
+//constructor
+AP_DEVO_Telem::AP_DEVO_Telem(AP_AHRS &ahrs, const AP_BattMonitor &battery) :
+    _ahrs(ahrs),
+    _battery(battery),
+    _port(NULL),
+    _initialised_uart(false),
+    _last_frame_ms(0)
+    {}
+
+// init - perform require initialisation including detecting which protocol to use
+void AP_DEVO_Telem::init(const AP_SerialManager& serial_manager)
+{
+    
+
+    // check for DEVO_DPort
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Devo_Telem, 0))) {
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        // initialise uart
+        _port->begin(AP_SERIALMANAGER_DEVO_TELEM_BAUD, AP_SERIALMANAGER_DEVO_BUFSIZE_RX, AP_SERIALMANAGER_DEVO_BUFSIZE_TX);
+        _initialised_uart = true;   // SerialManager initialises uart for us
+
+
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_DEVO_Telem::tick, void));
+    }
+}
+
+
+uint32_t AP_DEVO_Telem::gpsDdToDmsFormat(float ddm){
+    int32_t deg = (int32_t)ddm;
+    float mm = (ddm - deg) * 60.0f;
+
+    mm = ((float)deg * 100.0f + mm) /100.0;
+
+    if ((mm < -180.0f) || (mm > 180.0f)){
+        mm = 0.0f;
+    }
+
+    return mm * 1.0e7;
+}
+
+
+/*
+  send_frames - sends updates down telemetry link 
+  should be called by main program at 1hz
+*/
+
+#define DEVO_SPEED_FACTOR 0.0194384f
+
+void AP_DEVO_Telem::send_frames(uint8_t control_mode)
+{
+    // return immediately if not initialised
+    if (!_initialised_uart) {
+        return;
+    }
+
+    const AP_GPS &gps = AP::gps();
+
+    if (gps.status() >= 3) {
+        struct Location loc = gps.location();//get gps instance 0
+
+        devoPacket.lat = gpsDdToDmsFormat(loc.lat);
+        devoPacket.lon = gpsDdToDmsFormat(loc.lng);
+        devoPacket.speed = (int16_t)(gps.ground_speed() * DEVO_SPEED_FACTOR * 100.0f);  // * 100 for cm
+
+        if(_ahrs.get_position(loc)) {
+            int32_t baro_alt = loc.alt; // in cantimeters as DEVO needs
+            if (!loc.flags.relative_alt) {
+                baro_alt -= _ahrs.get_home().alt; // subtract home if set, in cantimeters too
+            }
+        /*
+          Note that this isn't actually barometric altitude, it is the
+          inertial nav estimate of altitude above home.
+        */
```

Comment:
What should this send? Altitude above home or absolute altitude?

Reply:
no one knows where which altitude is

Reply:
What? I clearly now where altitudes are... And I'm sure other people know too. I ask again, what is this supposed to send?

Reply:
When I asked OSD-related questions about "whch altitude comes in those MAVlink messages" I don"t got clear answer. This should be altitude above home.

Reply:
I don't know who you asked but it wasn't me. If it's altitude above home, you can skip all that code above with just: 
```
float alt;
ahrs.get_relative_position_D_home(alt);
alt *= -100.0f;
```

This also always returns a value, even if it's directly from baro.